lib/Doctrine/DBAL/Types/TypeRegistry.php

```diff
@@ -0,0 +1,118 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\DBAL\Types;
+
+use Doctrine\DBAL\DBALException;
+use function array_search;
+use function get_class;
+
+/**
+ * @internal
+ */
+final class TypeRegistry
+{
+    /**
+     * Map of already instantiated type objects. One instance per type (flyweight).
+     *
+     * @var Type[]
+     */
+    private $typeInstances = [];
+
+    /**
+     * Map of defined types and their classes
+     *
+     * @var string[]
+     */
+    private $typesMap = [];
+
+    /**
+     * Factory method to create type instances.
+     * Type instances are implemented as flyweights.
+     *
+     * @throws DBALException
+     */
+    public function getType(string $name) : Type
+    {
+        if (isset($this->typeInstances[$name])) {
+            return $this->typeInstances[$name];
+        }
+
+        if (isset($this->typesMap[$name])) {
+            return $this->typeInstances[$name] = new $this->typesMap[$name]();
+        }
+
+        throw DBALException::unknownColumnType($name);
+    }
+
+    /**
+     * Find a name for the given type.
+     *
+     * @throws DBALException
+     */
+    public function getName(Type $type) : string
+    {
+        $name = array_search($type, $this->typeInstances, true);
+
+        if ($name === false) {
+            throw DBALException::typeNotFound(get_class($type));
+        }
+
+        return $name;
+    }
+
+    /**
+     * Checks if there is a type of the given name.
+     */
+    public function hasType(string $name) : bool
+    {
+        return isset($this->typesMap[$name]);
+    }
+
+    /**
+     * Adds a custom type to the type map.
+     *
+     * @param string $name      The name of the type. This should correspond to what getName() returns.
+     * @param string $className The class name of the custom type.
+     *
+     * @throws DBALException
+     */
+    public function addType(string $name, string $className) : void
+    {
+        if (isset($this->typesMap[$name])) {
+            throw DBALException::typeExists($name);
+        }
+
+        $this->typesMap[$name] = $className;
+    }
+
+    /**
+     * Overrides an already defined type to use a different implementation.
+     *
+     * @throws DBALException
+     */
+    public function overrideType(string $name, string $className) : void
+    {
+        if (! isset($this->typesMap[$name])) {
+            throw DBALException::typeNotFound($name);
+        }
+
+        if (isset($this->typeInstances[$name])) {
+            unset($this->typeInstances[$name]);
+        }
+
+        $this->typesMap[$name] = $className;
+    }
+
+    /**
+     * Gets the types array map which holds all registered types and the corresponding
+     * type class
+     *
+     * @return string[]
+     */
+    public function getTypesMap() : array
```

Comment:
Besides BC, what could this method be used for? Unless it's by design, we should mark it `@internal`.

Given that we have two maps internally, and `getType()` returns an instance, would it make sense to name it as `getTypeClasses()`?

Reply:
Whole TypeRegistry is `@internal` now. But yes, this method is useless and only for compatibility with `Type`.